ocaml/xapi/import.ml

```diff
@@ -1838,6 +1838,49 @@ module PVS_Site : HandlerTools = struct
   let handle __context _ _ _ _ _ () = ()
 end
 
+module VTPM : HandlerTools = struct
+  (* We expect a VTPM to be part of a VM's meta data and not to exist
+     independently. Hence, we never look for the the VTPM to exist
+     already but always create it *)
+
+  type precheck_t = Import of vtpm'
+
+  (* Compare the state of the database with the metadata to be imported.
+     Returns a result which signals what we should do to import the
+     metadata. *)
+
+  (** expected case *)
+  let precheck __context _config _rpc _session _state obj =
+    let vtpm' = vtpm'_of_rpc obj.snapshot in
+    debug "%s: importing VTPM" __FUNCTION__ ;
+    Import vtpm'
+
+  (* Handle the result of the precheck function, but don't create any
+     database objects.  Add objects to the state table if necessary, to
+     keep track of what would have been imported.*)
+  let handle_dry_run __context _config _rpc _session state obj = function
+    | Import _vtpm' ->
+        debug "%s: importing VTPM" __FUNCTION__ ;
+        let ref = Ref.make () in
+        state.table <- (obj.cls, obj.id, Ref.string_of ref) :: state.table
+
+  (* Handle the result of the check function, creating database objects
+     if necessary.  For certain combinations of result and object type,
+     this can be aliased to handle_dry_run. *)
+  let handle __context _config _rpc _session state _obj = function
+    | Import vtpm' ->
+        debug "%s: importing VTPM" __FUNCTION__ ;
+        let vm =
+          log_reraise
+            ("Failed to find VTPM's VM: " ^ Ref.string_of vtpm'.vTPM'_VM)
+            (lookup vtpm'.vTPM'_VM) state.table
+        in
+        let self =
+          Xapi_vtpm.create ~__context ~vM:vm ~is_unique:vtpm'.vTPM'_is_unique
+        in
+        Xapi_vtpm.set_contents ~__context ~self ~contents:vtpm'.vTPM'_content
```

Comment:
There are a few more fields that are part of the incoming record. Shouldn't we set those?

Reply:
What about `backend`, `persistence_backend` and `uuid` (other handlers preserve the uuid if `config.full_restore`; see VIF and VBD)? The `protected` field may be hardcoded now, but I assume later won't be, which may lead to surprises.

Many handlers use `Client.X.create_from_record` to set all fields at once. You could do that as well, after reconstructing the actual datamodel record.

Reply:
can we set those fields to their defaults? doesn't matter what the backend was previously, we want to use the latest one. we do want to preserve the vtpm profile though.

Reply:
Is it safe to update the UUID? Currently we we get a new UUID but I would assume it could be likewise kept. The problem here is that this is: 
* if the UUID already exists, I probably should not import the VTPM and use the existing one - but sharing between VTPMs seems not like a good idea
* if I create a new VTPM but update the UUID, I could end up with a UUID that is not unique